### PR TITLE
Steam Family Sharing support

### PR DESCRIPTION
**Pull request by vrokolos**
Added support for steam family sharing with achievements and game time as well!!

In steam you can state which friends have shared games with you by changing their nicknames and adding "#" anywhere in it. Unfortunately I don't know of any automatic way to get them. If you do please notify me.

This adds all their games that you don't own to your gog library

#12

### CODE REVIEW

File: src/plugin.py [hunk 1]
```diff
@@ -76,6 +77,9 @@ def parse_stored_cookies(cookies):
 class SteamPlugin(Plugin):
     def __init__(self, reader, writer, token):
         super().__init__(Platform.Steam, __version__, reader, writer, token)
+        self._own_games: List = []
```
Comment (vrokolos) [inline src/plugin.py]: added some cache so we don't hit api every time

File: src/plugin.py [hunk 2]
```diff
@@ -202,11 +238,11 @@ def _force_utc(self):
     async def get_game_time(self, game_id: str, context: Any) -> GameTime:
         game_time = context.get(game_id)
         if game_time is None:
-            raise UnknownError("Game {} not owned".format(game_id))
+            logging.exception("Game {} not owned".format(game_id))
```
Comment (vrokolos) [inline src/plugin.py]: We don't want to stop if a game is not found

File: src/plugin.py [hunk 3]
```diff
@@ -202,11 +238,11 @@ def _force_utc(self):
...
         return game_time
 
     async def _get_game_times_dict(self) -> Dict[str, GameTime]:
-        games = await self._client.get_games(self._steam_id)
+        games = self._own_games
```
Comment (vrokolos) [inline src/plugin.py]: use cache instead of api

File: src/plugin.py [hunk 4]
```diff
@@ -191,8 +195,40 @@ def _force_utc(self):
...
+        newgames: List[Game] = []
+        self._other_games = []
+        for i in profiles:
+            othergames = await self._client.get_games(i.user_id)
+
+            try:
+                for game in othergames:
+                    hasit = any(f == str(game["appid"]) for f in owngames) or any(f.game_id == str(game["appid"]) for f in newgames)
+                    if not hasit:
+                        self._ither_games.append(str(game["appid"]))
```
Comment (marco-dasilva) [inline src/plugin.py]: Think it needs to change to `_other_games`

File: src/plugin.py [hunk 5]
```diff
@@ -226,6 +262,26 @@ def _force_utc(self):
             logging.exception("Can not parse backend response")
             raise UnknownBackendResponse()
 
+        try:
+            steamFolder = get_configuration_folder()
+            accountId = str( int(self._steam_id) - 76561197960265728  )
```
Comment (FriendsOfGalaxy) [inline src/plugin.py]: I assume this is steamid to steam community id conversion, i think this should be in a separate method then and the '76561197960265728 ' itself should be a const value under a variable

File: src/plugin.py [hunk 6]
```diff
@@ -226,6 +262,26 @@ def _force_utc(self):
...
+            vdfFile = os.path.join(steamFolder, "userdata", accountId, "config", "localconfig.vdf")
+            logging.debug(vdfFile)
```
Comment (FriendsOfGalaxy) [inline src/plugin.py]: `logging.debug(f"Users Localconfig.vdf {vdfFile}")`
or anything else which would help looking for this in the logs

File: src/plugin.py [hunk 7]
```diff
@@ -226,6 +262,26 @@ def _force_utc(self):
...
+            data = load_vdf(vdfFile)
+            timedata = data["UserLocalConfigStore"]["Software"]["Valve"]["Steam"]["Apps"]
+            for gameid in self._other_games:
+                playTime = 0
+                lastPlayed = 86400
```
Comment (FriendsOfGalaxy) [inline src/plugin.py]: Why 86400?
Add comment explaining or/and put in a variable

File: src/plugin.py [hunk 8]
```diff
@@ -191,8 +195,40 @@ def _force_utc(self):
...
+        
+        game_ids = list(map(lambda x: x.game_id, owned_games))
+        other_games = await self.get_steam_sharing_games(game_ids)
+        for i in other_games:
+            owned_games.append(i)
+
         return owned_games
 
+    async def get_steam_sharing_games(self,owngames: List[str]) -> List[Game]:
+        profiles = list(filter(lambda x: "!" in x.user_name, self._own_friends))
```
Comment (FriendsOfGalaxy) [inline src/plugin.py]: Add explaination why are we serching for "!"
Can a user create a username with "!" himself ?

File: src/plugin.py [hunk 9]
```diff
@@ -76,6 +77,9 @@ def parse_stored_cookies(cookies):
 class SteamPlugin(Plugin):
     def __init__(self, reader, writer, token):
         super().__init__(Platform.Steam, __version__, reader, writer, token)
+        self._own_games: List = []
+        self._other_games: List[str] = []
```
Comment (FriendsOfGalaxy) [inline src/plugin.py]: `self._family_sharing_games: List[str] = []`

File: src/plugin.py (hunk 7 above)
Comment (FriendsOfGalaxy) [reply src/plugin.py]: I see its used in code above it, we should put it in a var then

File: src/plugin.py [hunk 10]
```diff
@@ -226,6 +262,26 @@ def _force_utc(self):
             logging.exception("Can not parse backend response")
             raise UnknownBackendResponse()
 
+        try:
```
Comment (FriendsOfGalaxy) [inline src/plugin.py]: Is parsing local files the only solution for getting game times for family shared games? I guess its ok if there is no other way but i want to be sure we checked other possibilities first

File: src/plugin.py (hunk 3 above)
Comment (FriendsOfGalaxy) [reply src/plugin.py]: This method is called be prepare game times context which gets a list of ids from galaxy

if we want to use cache we should first check if all the games in the list of ids are present in the cache, if not we should hit steam api to be sure we have the freshest results.

**Comment by FriendsOfGalaxy** [comment]
Hey @vrokolos 

Thanks for the pull request! I've added some comments which should be resolved before we merge it, otherwise the idea is awesome.

File: src/plugin.py (hunk 8 above)
Comment (vrokolos) [reply src/plugin.py]: You can change the nickname of your friends in the steam client and include some pattern which I now have set as "!". I can't find a way to automatically get which friends have shared games with you so this is a workaround.

File: src/plugin.py (hunk 10 above)
Comment (vrokolos) [reply src/plugin.py]: Can't find other way sorry

File: src/plugin.py (hunk 8 above)
Comment (vrokolos) [reply src/plugin.py]: Ah I see what you mean. I wish there was plugin configuration on galaxy integration plugins so we can select and store the list of friends there.
What we can do is check for both ending with "*" (which means nickname) and "!"

**Comment by vrokolos** [comment]
So I changed the code added some of your suggestions and made it compatible with latest version.

**Comment by FriendsOfGalaxy** [comment]
Hi, we're reviewed your PR one more time and sadly have to reject it.

The first thing is that a user has to perform additional actions to get it work:
- rename his family sharing friend to add "#" at the end
- make sure his profile is not private

But as I said before - this is OK, as do not breaks anything.

Another problem is that many games are not shareable, and this PR just tries to fetch *all* friend games.

And finally there is a problem in Galaxy, that if one instance of your plugin returns different list of owned games than the other, user's games in local plugin context will be synchronized against each other: https://github.com/gogcom/galaxy-integrations-python-api/issues/60
And looks like there currently no workaround for this problem.

Additionally we will be switching to completely different way of fetching things from Steam, hopefully shared libraries will be there as well.

Anyway thanks for your contribution; we'll remember about this feature in future versions.

**Comment by INFINI7I** [comment]
How to add my second account? What I must to do?

**Pull request closed without merging**